scalanativelib/src/mill/scalanativelib/testinterface/ComRunner.scala

```diff
@@ -0,0 +1,128 @@
+
+  private[this] val runner = new Thread {
+    override def run(): Unit = {
+      val port = serverSocket.getLocalPort
+      val allArgs = Vector(bin.toString, port.toString) ++ args
+      println(s"Starting process $bin on port $port")
+      Jvm.baseInteractiveSubprocess(
+        allArgs,
+        envVars,
+        workingDir = ammonite.ops.pwd)
+    }
+  }
+
+  private[this] var serverSocket: ServerSocket = _
+  private[this] val socket =
+    try {
+      serverSocket = new ServerSocket(0)
+
+      runner.start()
+
+      serverSocket.setSoTimeout(40 * 1000)
+      serverSocket.accept()
+    } catch {
+      case _: SocketTimeoutException =>
+        throw new MessageOnlyException(
+          "The test program never connected to sbt.")
+    } finally {
+      // We can close it immediately, since we won't receive another connection.
+      serverSocket.close()
+    }
+
+  private[this] val in = new DataInputStream(
+    new BufferedInputStream(socket.getInputStream))
+  private[this] val out = new DataOutputStream(
+    new BufferedOutputStream(socket.getOutputStream))
+
+  /** Send message `msg` to the distant program. */
+  def send(msg: Message): Unit = synchronized {
+    try SerializedOutputStream(out)(_.writeMessage(msg))
+    catch {
+      case ex: Throwable =>
+        close()
+        throw ex
+    }
+  }
+
+  /** Wait for a message to arrive from the distant program. */
+  def receive(timeout: Duration = Duration.Inf): Message =
+    synchronized {
+      in.mark(Int.MaxValue)
+      val savedSoTimeout = socket.getSoTimeout()
+      try {
+        val deadLineMs = if (timeout.isFinite()) timeout.toMillis else 0L
+        socket.setSoTimeout((deadLineMs min Int.MaxValue).toInt)
+
+        val result =
+          SerializedInputStream.next(in)(_.readMessage()) match {
+            case logMsg: Log =>
+              log(logMsg)
+              receive(timeout)
+            case other =>
+              other
+          }
+
+        in.mark(0)
+        result
+
+      } catch {
+        case _: EOFException =>
+          close()
+          throw new MessageOnlyException(
+            s"EOF on connection with remote runner on port ${serverSocket.getLocalPort}")
+        case _: SocketTimeoutException =>
+          close()
+          throw new TimeoutException("Timeout expired")
+        case ex: Throwable =>
+          close()
+          throw ex
+      } finally {
+        socket.setSoTimeout(savedSoTimeout)
+      }
+    }
+
+  def close(): Unit = {
+    in.close()
+    out.close()
+    socket.close()
+  }
+
+  private def log(message: Log): Unit =
+    println(message)
+//    message.level match {
+//      case Level.Info  => logger.info(message.message)
+//      case Level.Warn  => logger.warn(message.message)
+//      case Level.Error => logger.error(message.message)
+//      case Level.Trace => message.throwable.foreach(logger.trace(_))
+//      case Level.Debug => logger.debug(message.message)
+//    }
```

Comment:
We have "no commented out code" policy in SN, is this left as an oversight?

Reply:
Just to remind me that the current situation is not ideal and it needs sorting out. Will be removed when the logging situation is resolved.

Thanks!